designs/2021-schema-object-rules/README.md

```diff
@@ -0,0 +1,284 @@
+- Remove handling of the function-style rule properties `rule.schema` and `rule.deprecated` in `createCoreRuleConfigs()` in `lib/init/config-rule.js`
+- Remove handling of the function-style rule property `rule.schema` in `getRuleOptionsSchema()` in `lib/config/rule-validator.js` and `lib/shared/config-validator.js`
+
+Test changes:
+
+- We'll change the test in `tests/lib/linter/rules.js` for function-style rules to ensure that function-style rules are NOT detected
+
+Some of these changes need to be replicated in `@eslint/eslintrc` as well.
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+
+### Documentation for requiring schemas
+
+The [Rule Basics](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics) section of the `Working with Rules` documentation page will be updated to mention when the `schema` property is required. We will also update any other documentation showcasing rules with options to include the `schema` property.
+
+### Documentation for requiring object-style rules
+
+We can finally delete `docs/developer-guide/working-with-rules-deprecated.md`, along with updating/removing any other documentation or test cases that use function-style rules.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+### Drawbacks of requiring schemas
+
+Plugin authors who left out schemas will need to take the time to add them.
+
+Users of old/unmaintained plugins/rules which are missing schemas will not be able to use them until they are updated.
+
+### Drawbacks of requiring object-style rules
+
+Plugin authors who still have function-style rules will need to convert them to object-style rules.
+
+Users of old/unmaintained plugins/rules which still have function-style rules will not be able to use them until they are updated.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+This RFC describes a set of breaking changes that can be released in the next major version of ESLint. Both ESLint plugin authors and ESLint users will be affected and will be unable to use offending rules until they are updated. This will mostly affect a small number of older ESLint plugins that haven't been updated nor maintained in years.
```

Comment:
This will be also a breaking change for end users who have invalid configurations for non-offending rules (those that don't expect any options), as they'll have to remove useless options from config files.

Reply:
I added a bullet point to denote this.